recipe/2024-01-support-private-terraform-repository.md

```diff
@@ -0,0 +1,237 @@
+
+Personal Access Token with Username way of authentication is supported by git from most of the platforms(e.g Bitbucket,Gitlab,Github etc) and generic git repository URL format as template path can be used to while registering recipe to access the terraform modules from private git repository from any platform.
+
+Generic Git Repository URL format:
+```diff
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'corerp-resources-recipe-env'
+  location: location
+  properties: {
+    ...,
+    recipes: {
+      'Applications.Datastores/mongoDatabases':{
+        recipe1: {
+          templateKind: 'terraform'
++          templatePath: "git::https://{username}:{PERSONAL_ACCESS_TOKEN}@example.com.com/test-private-repo.git"
+        }
+      }
+    }
+  }
+}
+```
+this takes latest as the default version if not provided. To specify a particular version the URL should be appended with `ref=<version>"`
+E.g:
+```
+templatePath: "git::https://{username}:{PERSONAL_ACCESS_TOKEN}@example.com.com/test-private-repo.git?ref=v1.2.0"
+```
+
+Since adding sensitive information like tokens to the terraform configuration files which may be stored in version control can pose security issues. So we could use different ways to store the github credentials.
+
+Option 1 : Using Git Credential Store
+
+Git needs to know when and where to use the token when checking out code from a private repository. We want git to automatically detect when Terraform modules are being loaded from a private repository and insert the token for the duration of the session. This can be done by running the below command.
+```
+git config --global url."https://<username>:<personal-access-token>@dev.azure.com.com".insteadOf https://dev.azure.com
+```
+or by adding this entry on the .gitconfig file on the cluster.
+```
+[url "https://<username>:<personal-access-token>@dev.azure.com.com""]
+	insteadOf = https://dev.azure.com
+```
+But it wont work well when we have multiple environments. As global .gitconfig file has the git credentials stored for all the environments and the module repositories may be from different accounts.
+
+Option 2: Saving it as part of the environment resource.
+
+Add a new property `recipeConfig` write-only to store the information about the git credentials, and have a custom action `getRecipeConfiglike` (like `list-secrets`) to get these details. 
+
+```diff
++@secure()
++param username string
++@secure()
++param token string
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'dsrp-resources-env-recipes-context-env'
+  location: 'global'
+  properties: {
+    compute: {
+      ...
+    }
+    providers: {
+      ...
+    }
++   recipeConfig: {
++     terraform: {
++       secrets: {
++         "dev.azure.com": {
++            username : username
++            pat: token
```

Comment:
Are there potential security issues here if usernames and PAT are being passed through the bicep templates?

Reply:
If they're an input parameter to the Bicep template they will need to be marked/decorated `@secure()`

```bicep
@secure()
param token string
```

This tells Bicep/ARM not to log/store the value.

And then it will be up to the resource provider to securely store the value and make it a read-only API with a separate way to access its value. Same way we handle connection strings and passwords for portable resources.